include/wil/cppwinrt_authoring.h

```diff
@@ -0,0 +1,236 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+
+namespace wil
+{
+#ifndef __WIL_CPPWINRT_AUTHORING_INCLUDED
+#define __WIL_CPPWINRT_AUTHORING_INCLUDED
+    namespace details
+    {
+        template<typename T>
+        struct single_threaded_property_storage
+        {
+            T m_value{};
+            single_threaded_property_storage() = default;
+            single_threaded_property_storage(const T& value) : m_value(value) {}
+            operator T& () { return m_value; }
+            operator T const& () const { return m_value; }
+            template<typename Q> auto operator=(Q&& q)
+            {
+                m_value = wistd::forward<Q>(q);
+                return *this;
+            }
+        };
+    }
+
+    template <typename T>
+    struct single_threaded_ro_property : std::conditional_t<std::is_scalar_v<T>, wil::details::single_threaded_property_storage<T>, T>
+    {
+        single_threaded_ro_property() = default;
+        single_threaded_ro_property(const T& t) : base_type(t) { }
+        
+        const auto& operator()()
+        {
+            return *this;
+        }
+        template<typename Q> auto operator=(Q&& q) = delete;
+    private:
+        using base_type = std::conditional_t<std::is_scalar_v<T>, wil::details::single_threaded_property_storage<T>, T>;
+    };
+
+    template <typename T>
+    struct single_threaded_rw_property : single_threaded_ro_property<T>
+    {
+        template<typename... TArgs>
+        single_threaded_rw_property(TArgs&&... args) : single_threaded_ro_property<T>(std::forward<TArgs>(args)...) { }
+
+        using single_threaded_ro_property<T>::operator();
+
+        auto& operator()()
+        {
+            return *this;
+        }
+        template<typename Q> auto& operator()(Q&& q)
+        {
+            *this = std::forward<Q>(q);
+            return *this;
+        }
+
+
+        template<typename Q> auto operator=(Q&& q)
+        {
+            return static_cast<T&>(*this) = std::forward<Q>(q);
+        }
+    };
+
+#endif // __WIL_CPPWINRT_AUTHORING_INCLUDED
+
+#if !defined(__WIL_CPPWINRT_AUTHORING_INCLUDED_FOUNDATION) && defined(WINRT_Windows_Foundation_H) // WinRT / XAML helpers
+#define __WIL_CPPWINRT_AUTHORING_INCLUDED_FOUNDATION
+    namespace details {
+        template<typename T>
+        struct event_base {
+            winrt::event_token operator()(T&& handler) {
+                return m_handler.add(std::forward<T>(handler));
+            }
+            auto operator()(const winrt::event_token& token) noexcept { return m_handler.remove(token); }

```

Comment:
Does `winrt::event` not throw if the token is not found?

Reply:
no, it is a no-op if it's the wrong token.